test/control-plane: Add tests for service load-balancing

joamaki opened the pull request:
This adds control-plane tests for service load-balancing, e.g. tests
that inject k8s objects into the agent and verify the datapath state
changes accordingly.

The input k8s objects are generated (manually) with a script that boots up a kind
cluster, applies configuration to it and finally dumps the cluster state
that we exercise the agent with. These yaml files are then fed to the agent
using the fake k8s clients.

This makes progress towards #20160, with the tests here mapping to existing ginkgo e2e tests:

* graceful-termination: "Checks graceful termination of service endpoints"
* nodeport: "Checks E/W loadbalancing / Checks in-cluster KPR / Tests NodePort",
                  "Checks E/W loadbalancing / Checks in-cluster KPR / Tests HostPort"
* dual-stack: "Checks E/W loadbalancing / Checks DualStack services"

joamaki committed:
pkg/k8s/slim: Regenerate with proper groupNames

The generated fake clients had wrong group names, which caused
schema mismatches. Add the correct groupName annotations.
Note that core doesn't have a group name, but it needs "groupGoName"
so it will do a named import that doesn't clash.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
k8s: Drop use of RESTClient in favour of typed client

By constructing the ListerWatcher from the code-generated typed K8s
client, we can use the generated fake clients to inject events to the
agent for testing purposes.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
lbmap: Add LBMap interface and move lbmap construction to datapath

To make testing of the service package simpler, move the LBMap
under the datapath interface. This simplifies the injection of
the MockLBMap.

Add Name and Namespace to UpsertServiceParams to make writing test
assertions against MockLBMap easier.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
daemon: Add few missing DryMode checks

Add checks for DryMode to allow writing a control-plane test
for services.

Remove skipInit() so that the setup for the flags isn't skipped.
Without this, option.Config.Populate would fail as the defaults
would not be filled in. Removing this had no effect on tests,
so should be safe to remove.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

christarazi commented on pkg/k8s/utils/listwatcher.go:
```diff
@@ -0,0 +1,70 @@
+// which would not be possible if we used NewListWatchFromClient and RESTClient().
+func ListerWatcherFromTyped[T k8sRuntime.Object](lw typedListWatcher[T]) cache.ListerWatcher {
+	return &genListWatcher[T]{ctx: context.TODO(), lw: lw}
+}
```
This should probably take in `context.Context` so we can properly pass it from the caller rather than hardcode to `context.TODO()` now. Plus we might as well do this now before this becomes an API that others depend on so it's better to change the API now to account for `context.Context`. WDYT?

joamaki replied:
Hm, this context should be per-request, so perhaps better to move the context creation into the List and Watch wrappers. We don't really have a context available in any of the places where we'd want to use ListerWatcherFromTyped so not sure it's worth exposing it as we'd need to do fair bit of refactoring (unless we just call `context.Background()` every time). OTOH, it would be a better API to be able to cancel the list or watch requests, but it's not something we had before.

Here's btw what client-go does (https://github.com/kubernetes/client-go/blob/master/tools/cache/listwatch.go#L87) with the ListWatchFromClient that we used earlier:
```
func NewFilteredListWatchFromClient(c Getter, resource string, namespace string, optionsModifier func(options *metav1.ListOptions)) *ListWatch {
	listFunc := func(options metav1.ListOptions) (runtime.Object, error) {
		optionsModifier(&options)
		return c.Get().
			Namespace(namespace).
			Resource(resource).
			VersionedParams(&options, metav1.ParameterCodec).
			Do(context.TODO()).
			Get()
	}
```

:smile:

christarazi replied:
Interesting, well if client-go isn't doing context plumbing then I guess we don't need to either. I am surprised though. :shrug:

christarazi commented on pkg/k8s/version/version.go:
```diff
@@ -117,6 +117,9 @@ func updateVersion(version semver.Version) {
 	cached.capabilities.EndpointSliceV1 = isGEThanAPIDiscoveryV1(version)
+	if cached.capabilities.EndpointSliceV1 {
+		cached.capabilities.EndpointSlice = true
+	}
```
This feels like it should be in its own commit

joamaki replied:
Agreed, will move it.

aanm replied:
It looks that this is kind of surfacing a bug introduced by https://github.com/cilium/cilium/commit/7a1039f4143eb6664a379996264127efdfb4f638 where the assumption made before that commit was that `cached.capabilities.EndpointSlice` could be false but `cached.capabilities.EndpointSliceV1` wasn't. @joamaki this also needs to be in a separate PR so it can be backported to all branches.

christarazi commented on pkg/k8s/version/version.go:
```diff
@@ -297,6 +300,9 @@ func Update(client kubernetes.Interface, conf k8sconfig.Configuration) error {
 		// groups will only become discoverable once Cilium is up.
 		_, apiResourceLists, err := client.Discovery().ServerGroupsAndResources()
+
+		fmt.Printf("apiResourceLists: %v\n", apiResourceLists)
```
Leftover?

christarazi commented on pkg/k8s/client.go:
```diff
@@ -42,7 +42,7 @@ var (
 
 	// k8sWatcherCLI is the client dedicated k8s structure watchers.
-	k8sWatcherCLI = &K8sClient{}
+	k8sWatcherCLI = &K8sSlimClient{}
```
How does this work if this client is also used for other K8s resoruces watches that aren't on slim objects? Or is another client used for non-slim objects?

joamaki replied:
The `K8sCLI` is for the non-slim objects. `K8sWatcherCLI` only has the generated typed clients for the slim objects, but its `RESTClient()` probably would work for non-slim core objects.

christarazi replied:
> its RESTClient() probably would work for non-slim core objects.

But isn't that problematic because it would fetch the slim version of the core objects?

joamaki replied:
My understanding is that the `RESTClient()` is not tied to the scheme, but it's just the generic client, and for the slim client the API path will be the same.

Note that the "slim objects" are just the regular k8s objects, but with fewer fields defined so that we don't unmarshal and store fields that are unrelevant to cilium.

christarazi commented on pkg/datapath/linux/datapath.go:
```diff
@@ -27,17 +29,23 @@ type linuxDatapath struct {
 
 // NewDatapath creates a new Linux datapath
 func NewDatapath(cfg DatapathConfiguration, ruleManager datapath.IptablesManager, wgAgent datapath.WireguardAgent) datapath.Datapath {
+
```
Nit: no need for extra line

christarazi commented on test/control-plane/services/dual-stack/generate.sh:
```diff
@@ -0,0 +1,47 @@
+	manifests/echo_svc_dualstack.yaml
+)
+
+: Start a kind cluster
```
Why are these not just comments with `#`?

joamaki replied:
This is a pattern used with `set -x` to show "comments" on what's happening. Picked this up long time ago from somewhere. E.g. it's like "echo" but without so much noise in the xtrace output.

joamaki replied:
```
/tmp $ cat test.sh
#!/usr/bin/env bash
set -x

echo "hello"
: this is a : comment
# this is a # comment

/tmp $ ./test.sh
+ echo hello
hello
+ : this is a : comment
```

christarazi replied:
:+1:

christarazi commented on test/control-plane/services/nodeport/README:
```diff
@@ -0,0 +1,8 @@
+
```
Nit: extra line and file should probably have an extension like `.md`?

christarazi commented on test/control-plane/services/runner.go:
```diff
@@ -0,0 +1,463 @@
+// Run sets up the control-plane with a mock lbmap and executes the test case
+// against it.
+func (testCase *ServicesTestCase) Run(t *testing.T, modConfig func(*option.DaemonConfig)) {
+
```
Nit: extra line here too

christarazi commented on test/control-plane/services/runner.go:
```diff
@@ -0,0 +1,463 @@
+		}
+		testCase.InitialObjects = unmarshalList(bs)
+	}
+
```
What happens if `os.Stat` fails?

joamaki replied:
Then the set of initial objects is empty, which will cause the agent to fail to start as it can't find the `Node` resource for the local node. This wasn't the case in earlier iteration, but now that it is, I'll change this to fail here.

christarazi commented on test/control-plane/services/runner.go:
```diff
@@ -0,0 +1,463 @@
+	tempDir string
+}
+
+func startCiliumAgent(nodeName string, clients fakeClients, modConfig func(*option.DaemonConfig)) (datapath.Datapath, agentHandle) {
```
Nit: maybe it should be called `mustStartCiliumAgent` given that we expect it to panic when something goes wrong?

joamaki replied:
Changed it to return err and t.Fatal on it.

christarazi reviewed (changes requested):
Amazing start! :rocket: 

Some comments below. 

A couple of things that weren't clear to me:
* how do I run these?
* how do I add new tests?
* how do I parse the results?
* there's some notion of generating files, when do I need to do this?

 I think we'll need to document the above on docs.cilium.io and/or in the README of `test/control-plane` like you added. We can probably add these things in later PRs also.

aditighag commented on pkg/maps/lbmap/lbmap.go:
```diff
@@ -79,6 +79,8 @@ type UpsertServiceParams struct {
 	UseMaglev                 bool
 	L7LBProxyPort             uint16 // Non-zero for L7 LB services
+	Name                      string // Name of the k8s Service
+	Namespace                 string // Namespace of the k8s Service
```
Use the `ServiceID` type from k8s package - https://github.com/cilium/cilium/blob/master/pkg/k8s/service.go#L242.

joamaki replied:
IMO we shouldn't be depending on the k8s package from here. Perhaps ServiceID could be moved to e.g. `pkg/loadbalancer`?

aditighag replied:
Yes, let's move it to `pkg/loadbalancer` or `pkg/service`.

joamaki replied:
`pkg/loadbalancer` already has `type ServiceID uint16`. Perhaps we can come up with another name for this. Ok if we'll do this in follow-up?

joamaki replied:
Ended up moving `type Name struct` from `pkg/service` to `pkg/loadbalancer` as `type ServiceName struct` and refactoring the code to use it. Even though it currently is the k8s definition of a service name I think it makes sense to keep it separate since loadbalancing can be used outside k8s and we might want to tweak how we deal with names in the future.

aditighag commented on test/control-plane/README.md:
```diff
@@ -0,0 +1,51 @@
+is asynchronous, the validation is repeated until either timeout is reached or
+it passes.
+
+Tests can be written either by manually defining the objects and the validation
+of each step or as a golden test case where the input objects are unmarshalled
+from YAML files.
```
>Tests can be written either by manually defining the objects and the validation
of each step 

Is there an example for writing a test this way?

aanm replied:
Writing a comment here because I'm also interested in the answer. Or at least, can it be a seperate commit where a new tests is added?

joamaki replied:
Heh, should've known I won't get a way with just a remark about it :D. I'll write a "minimal" example.

joamaki replied:
See `test/control-plane/example/minimal_test.go`. Maybe not so much an example actually so should perhaps rename. Suggestions?

aditighag commented on daemon/cmd/daemon_main.go:
```diff
@@ -147,6 +147,10 @@ func Execute() {
 }
 
 func skipInit(basePath string) bool {
+	if strings.HasSuffix(basePath, ".test") {
```
Can we fold this check to the switch case below?

joamaki replied:
Hm, now this function will always return false since the only way to get here is via either the `cilium-agent` binary or a test binary (with suffix ".test"). Nothing breaks when I remove it, so I assume it's no longer necessary and can be removed.

EDIT: Double-checked and the commit (b5f1d9f2f0d2293f15fdb1f11f4275be28b47807) that introduced skipInit() said it's because cilium agent cli imports it, but this is no longer the case. Only places that import daemon/cmd are daemon/main.go and test/control-plane/agent.go.

aditighag commented on daemon/cmd/daemon.go:
```diff
@@ -401,10 +401,13 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
-		return nil, nil, fmt.Errorf("unable to initialize Kube proxy replacement options: %w", err)
+	var isKubeProxyReplacementStrict bool
+	if !option.Config.DryMode {
+		isKubeProxyReplacementStrict, err = initKubeProxyReplacementOptions()
```
We needn't skip the initialization of kpr options altogether - just the probing part.

joamaki replied:
Yep I should've been more careful and not leave the result of exploration in. I'll remove this and make sure initKubeProxyReplacement() works in DryMode (e.g. no probing).

aditighag commented on pkg/datapath/linux/datapath.go:
```diff
@@ -27,17 +29,22 @@ type linuxDatapath struct {
 // NewDatapath creates a new Linux datapath
 func NewDatapath(cfg DatapathConfiguration, ruleManager datapath.IptablesManager, wgAgent datapath.WireguardAgent) datapath.Datapath {
+	maglev := option.Config.NodePortAlg == option.NodePortAlgMaglev
+	maglevTableSize := option.Config.MaglevTableSize
```
We are leaking service internal configs to `pkg/datapath`. This should be in the lbmap constructor.

joamaki replied:
Fixed.

aditighag reviewed (changes requested):
Thanks for tackling this! 🚀

aanm commented on pkg/k8s/slim/k8s/api/core/v1/doc.go:
```diff
@@ -5,7 +5,7 @@
 // +k8s:deepcopy-gen=package
```
This commit could be in a different PR and potentially backported to all branches.

joamaki replied:
If it's ok with you, I'd backport this commit by hand if/when we need to do it.

aanm commented on pkg/k8s/annotate.go:
```diff
@@ -31,6 +32,11 @@ type K8sClient struct {
 	kubernetes.Interface
 }
 
+// K8sSlimClient is a wrapper around slimclientset.Clientset.
+type K8sSlimClient struct {
+	slimclientset.Interface
+}
+
```
Out of curiosity, why do we need this wrapper?

joamaki replied:
I don't know why all the clients are wrapped like this in structs in the k8s package. Should I remove them? Only `K8sClient` has methods on it, rest (K8sCiliumCLI, K8sWatcherCLI, K8sAPIExtCli, K8sAPIExtWatcherCLI) are just pure wrappers.

aanm commented on pkg/k8s/init.go:
```diff
@@ -189,7 +189,7 @@ func Init(conf k8sconfig.Configuration) error {
 	}
 
-	if option.Config.K8sHeartbeatTimeout != 0 {
+	if !option.Config.DryMode && option.Config.K8sHeartbeatTimeout != 0 {
```
Please add a comment with the "why?" If you are adding this `option.Config.DryMode` does it mean that it is not possible to unit test the heartbeat logic?

joamaki replied:
Ah right now `Init` isn't even called by these tests, so this change is unnecessary. Will drop it.

It is possible to do health checks against the fake clients, as long as it goes via the typed client and not raw RESTClient.

aanm commented on pkg/maps/lbmap/types.go:
```diff
@@ -11,6 +11,24 @@ import (
+	DumpBackendMaps() ([]*loadbalancer.Backend, error)
+	DumpAffinityMatches() (BackendIDByServiceIDSet, error)
+	DumpSourceRanges(bool) (SourceRangeSetByServiceID, error)
+}
```
Can you explain this move? To me it's odd to see an interface being part of the package where its implementation lives. Shouldn't an interface be defined somewhere else? See [1] to continue.

aanm commented on pkg/service/service.go:
```diff
@@ -37,24 +37,6 @@ var (
-	DumpAffinityMatches() (lbmap.BackendIDByServiceIDSet, error)
-	DumpSourceRanges(bool) (lbmap.SourceRangeSetByServiceID, error)
-}
-
```
[1] The idea was that `pkg/service` wouldn't need to import the `lbmap` package. But I guess that cat is out of the box so you can disregard my comment of interfaces living in a package outside of their implementation.

joamaki replied:
I wanted to move the LBMap ownership to datapath, so I needed the interface out from `pkg/service`. Initially I moved the LBMap interface to `pkg/datapath/types` but that felt silly as we had `pkg/maps/lbmap`. But I see your point. Perhaps this should be in `pkg/datapath/types` instead and then we'd have the concrete implementations in `pkg/maps/lbmap` and `pkg/testutils/mockmaps`. (longer term I'd like these to move under `pkg/datapath` as they're implementation details of the datapath).

Would `pkg/datapath/types` seem like a reasonable place for `LBMap` as it's logically part of the `Datapath` interface?

aanm replied:
> Would `pkg/datapath/types` seem like a reasonable place for `LBMap` as it's logically part of the `Datapath` interface?

@joamaki yes, I think that's perfect.

joamaki replied:
Done. I also of course had to move `UpsertServiceParams` and few others there.

aanm commented on test/control-plane/README.md:
```diff
@@ -0,0 +1,51 @@
+If the test case is a golden test, the golden output files can be updated
+with the '-update' flag:
+
+  $ go test ./test/control-plane/services/dual-stack -test.v -update
```
Can we make this a make target? I'll never remember this specially if the number test features grow :see_no_evil:

joamaki replied:
For these tests it makes more sense to run directly via `go test` as you often want to set `-test.v` or `-debug` etc. and we don't have make targets for other go tests either, so I would argue against this, but don't have a strong opinion.

aanm replied:
But we do have a `make integration-tests`, so I was thinking a `make tests` that would be a wrapper for `go test ./... -test.v`  and `make update-tests` for `go test ./... -update`

joamaki replied:
Ah sure. Having it in `test/control-plane/Makefile` as kind of a cheatsheet sounds good to me. Adding it.

aanm commented on test/control-plane/agent.go:
```diff
@@ -0,0 +1,115 @@
+	option.Config.K8sEnableK8sEndpointSlice = true
+	option.Config.EnableL7Proxy = false
+	option.Config.EnableHealthCheckNodePort = false
+	option.Config.Debug = true
```
We have all of these options, that's fine. Now, how can I run my tests if they depend on `EnableIPv6` to be `true` for example?

aanm commented on test/control-plane/services/dual-stack/generate.sh:
```diff
@@ -0,0 +1,47 @@
+)
+
+: Start a kind cluster
+kind create cluster --config kind-config.yaml --name dual-stack
```
Where can we specify the k8s version?

joamaki replied:
Good that you pointed this out. Especially for the service load-balancing we should have variants for the different k8s major versions due to the Endpoints/EndpointSlices stuff.

Kubernetes version can be set in kind config by using a specific image:
https://kind.sigs.k8s.io/docs/user/configuration/#kubernetes-version,
e.g. `image: kindest/node~v1.21.12` etc.

I'll take a look adding tests for different versions and hard-coding
the version in the generation scripts.

joamaki replied:
Ok, now nodeport and dual-stack have per-version tests and all kind-config's hard-code the version. I didn't add more versions for graceful-termination as that relies on a newer feature and there were no differences in k8s inputs.

aanm commented on test/control-plane/services/dual-stack/dualstack_test.go:
```diff
@@ -0,0 +1,52 @@
+// SPDX-License-Identifier: Apache-2.0
```
Can we have these files `test/control-plane/services/dual-stack` to be part of a separate commit?

joamaki replied:
Done.

aanm commented on test/control-plane/services/dual-stack/lbmap1.golden:
```diff
@@ -0,0 +1,36 @@
+ - Services -------------------------------------------------------------------------
```
This is the expected output for a particular k8s versions, should we have different lbmap1.golden files depending on the k8s version? Probably it could be a symlinks pointing to the same output and if one of them differs then it wouldn't be a symlink.

joamaki replied:
Done.

aanm commented on test/control-plane/services/dual-stack/manifests/echo_svc_dualstack.yaml:
```diff
@@ -0,0 +1,19 @@
+apiVersion: v1
```
let's be consistent in the file names, let's change `echo-svc-dualstack` instead of `_`

joamaki replied:
should I also fix the source of this, e.g. `test/k8s/manifests/echo_svc_dualstack.yaml` and the other files there?

Note that I copied this and other manifests as slight tweaks were needed, and to me made sense to keep them with the specific test case.

aanm replied:
> should I also fix the source of this, e.g. `test/k8s/manifests/echo_svc_dualstack.yaml` and the other files there?
> 

Those will eventually disappear so there is no need to fix them.

> Note that I copied this and other manifests as slight tweaks were needed, and to me made sense to keep them with the specific test case.

joamaki replied:
Ok, fixed this one.

aanm commented on test/control-plane/services/dual-stack/state1.yaml:
```diff
@@ -0,0 +1,198 @@
+apiVersion: v1
```
what does `state1` filename means? Could we perhaps give a more descriptive filename?

joamaki replied:
This used to be events1.yaml, etc. but since we also need to do deletes I ended up making these represent the whole state (not including objects created in init.yaml), hence the "stateN.yaml" naming. I was thinking it'd be cleaner to do deletes explicitly,  but didn't yet have time to do that. I'll look into this and the non-golden test example today.

aanm commented on test/control-plane/services/dual-stack/generate.sh:
```diff
@@ -0,0 +1,47 @@
+kubectl wait --for=condition=ready --timeout=60s --all pods
+
+: Dump the services and endpoints
+kubectl get services,endpointslices -o yaml > events1.yaml
```
oh `events1.yaml` means this

aanm commented on test/control-plane/services/dual-stack/dualstack_test.go:
```diff
@@ -0,0 +1,52 @@
+	})
+
+	modConfig := func(c *option.DaemonConfig) {
+		c.EnableIPv6 = true
```
ah ah... that's how we modify specific options. Got it...

aanm commented on test/control-plane/services/dual-stack/dualstack_test.go:
```diff
@@ -0,0 +1,52 @@
+// SPDX-License-Identifier: Apache-2.0
```
Can we move all of the files that are **not** `_test.go` into a separate directory? We already have a `manifests` directory so it would make sense to have a directory for these input/output test files.

joamaki replied:
The `_test.go` is just one file with few lines on it. The test inputs and outputs are what matter in this directory so I at least would find it odd to move them into a sub-directory. 

For manifests it made sense as they together with generate.sh create the test input files and are not relevant most of the time. Moving kind-config.yaml and generate.sh to be together with manifests makes sense and cleans it up a bit. WDYT about having `generate` directory with all these in it and just have the `foo_test.go` along with test inputs and outputs in the test directory?

aanm replied:
> WDYT about having generate directory with all these in it and just have the foo_test.go along with test inputs and outputs in the test directory?

I think it's kind of fine but then it means we will end up with N input files where N is the k8s version. That's why I was thinking that it would make sense to have them in a separate directory as it is likely they will grow in number.

joamaki replied:
Well, right now every k8s version will need its own test binary and its own directory, so that kind of takes care of that concern. 

We'll need to revisit this question when we can call `NewDaemon` multiple times within one process. There I think organizing them into k8s version directories makes sense.

aanm commented on test/control-plane/services/graceful-termination/generate.sh:
```diff
@@ -0,0 +1,55 @@
+#!/usr/bin/env bash
```
Please make it in a separate commit.

joamaki replied:
Done.

aanm commented on test/control-plane/services/nodeport/README.md:
```diff
@@ -0,0 +1,8 @@
+
```
please put in a separate commit

joamaki replied:
Done.

aanm reviewed (changes requested):
Incredible PR :rocket:

joamaki committed:
test/control-plane: Add test framework for control-plane tests

Control-plane tests are tests that inject K8s objects into the cilium-agent
and validate the state of the (fake) datapath afterwards.

The tests can be written either as golden tests, e.g. where both input K8s
objects and the datapath state are expressed as files on disk (YAML), or
as tests with the inputs and validation expressed as Go code.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
test/control-plane: Add lbmap assertions and validation

This commit adds 'LBMapAssert', a helper for writing assertions
about the MockLBMap state and 'NewGoldenServicesTest', a golden
test suite with input k8s objects and output as lbmap serialized
in table form.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
test/control-plane: Add dual-stack test case

Test-case for checking that LBMap is correctly populated for both IPv4 and IPv6
services.

The test inputs are generated by the script 'generate.sh' that boots up a kind
cluster with a dual-stack configuration and starts a simple echo server.
The endpoints and services in that cluster state are then dumped to 'state1.yaml',
which then results in LBMap state as described by 'lbmap1.golden'.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
test/control-plane: Add graceful termination test

Test for graceful termination created with kind cluster that
has the graceful termination feature gate enabled. The generate.sh
script dumps the immediate states:
- state1.yaml: Server and client created and ready
- state2.yaml: Server pod marked to be deleted, backend becomes terminating
- state3.yaml: Server pod is gone

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
test/control-plane: Add tests for nodeport and hostport

Test case for checking that lbmap entries for nodeport and hostport
are correctly created. Test inputs generated by generate.sh that
creates a simple echo server that is exposed via ClusterIP, NodePort and
HostPort services.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

brb commented on test/control-plane/services/graceful-termination/graceful_termination_test.go:
```diff
@@ -0,0 +1,18 @@
+	modConfig := func(c *option.DaemonConfig) {
+		c.EnableK8sTerminatingEndpoint = true
+	}
+	NewGoldenServicesTest(t, "graceful-term-control-plane").Run(t, modConfig)
```
A comment saying what each state/golden corresponds to and what the golden test is testing would be useful.. Otherwise, a reader would need to read generate.sh.

joamaki replied:
Done.

brb reviewed (approved):
Amazing!

One comment + what others said.

BTW, are you going to remove the existing ginko test cases which you have implemented?

joamaki committed:
test/control-plane: Add test for node handler

This adds a minimal non-golden test case to validate that the datapath
node handler is called correctly. This also serves as an example of how
tests can be constructed from k8s objects defined in Go and with manual
validation functions.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

rolinh reviewed (approved):
Vendor changes lgtm

aditighag commented on test/control-plane/README.md:
```diff
@@ -0,0 +1,51 @@
+
+# Control-plane tests for the cilium agent
```
I'm missing some details around the link between installed cilium binary to validate k8s objects, and the mocked datapath maps. A diagram or some more details in the README could be useful (can be done in a follow-up PR).

joamaki replied:
This does not interact with a cilium binary. It calls `NewDaemon` directly, similar to what tests in `daemon/cmd` do.
My long-term vision for this is that these tests would only bring up the components they need rather than everything via NewDaemon. This of course pending on the modularization work...

tommyp1ckles commented on pkg/datapath/fake/datapath.go:
```diff
@@ -10,90 +10,103 @@ import (
 )
 
-type fakeDatapath struct {
+var _ datapath.Datapath = (*FakeDatapath)(nil)
```
What does this line do?

aditighag replied:
I would say it's meant to ensure that the type `FakeDatapath` implements the interface.

joamaki replied:
Yep, exactly. Asserts that `FakeDatapath` implements `datapath.Datapath`. https://go.dev/doc/effective_go#blank_implements.

tommyp1ckles commented on daemon/cmd/kube_proxy_replacement.go:
```diff
@@ -216,19 +219,22 @@ func initKubeProxyReplacementOptions() (bool, error) {
-		if h := probesManager.GetHelpers("sched_act"); h != nil {
-			if _, ok := h["bpf_fib_lookup"]; ok {
-				found = true
+		if probesManager != nil {
```
nit: just checking if we're not on dry mode here might make this easier for someone to read.

joamaki replied:
Fixed.

tommyp1ckles commented on test/control-plane/agent.go:
```diff
@@ -0,0 +1,118 @@
+func setupTestDirectories() string {
+	tempDir, err := ioutil.TempDir("", "cilium-test-")
+	if err != nil {
+		panic("TempDir() failed.")
```
nit: maybe append the error string to the panic message to give context to someone debugging.

joamaki replied:
Fixed.

tommyp1ckles commented on test/control-plane/marshalling.go:
```diff
@@ -0,0 +1,97 @@
+	var obj unstructured.Unstructured
+	err := yaml.Unmarshal([]byte(in), &obj)
+	if err != nil {
+		panic(err)
```
Just curious, why did you decide to have failures here panic? Think it would be good to have some context as to where this is being called from when panicing to help someone debugging a test.

tommyp1ckles replied:
Same for all of these

joamaki replied:
Mostly use of `panic` was because this was initially rather experimental, and as the stack trace with the error should've had enough context. But now that I look at this, it's much better that errors are bubbled up via `t.Fatalf` instead, especially if later on we'll be able to have multiple test cases in one binary. I'll clean all these up.

tommyp1ckles commented on test/control-plane/services/lbmap_validation.go:
```diff
@@ -0,0 +1,222 @@
+func newGoldenLBMapValidator(stateFile string, update bool) controlplane.Validator {
+	var v goldenLBMapValidator
+	var stepNum int
+	fmt.Sscanf(path.Base(stateFile), "state%d.yaml", &stepNum)
```
should we check for errors from Sscanf? might cause some confusing errors down the line if it fails silently.

joamaki replied:
Added the check.

tommyp1ckles commented on test/control-plane/services/lbmap_validation.go:
```diff
@@ -0,0 +1,222 @@
+		writeLBMapAsTable(&buf, lbmap)
+		if diff, ok := v.diffStrings(string(bs), buf.String()); !ok {
+			if v.update {
+				//t.Logf("lbmap mismatch:\n%s", diff)
```
Remove commented out code

joamaki replied:
Removed.

tommyp1ckles commented on test/control-plane/services/lbmap_validation.go:
```diff
@@ -0,0 +1,222 @@
+	} else {
+		// Mark failed as the expected output was missing, but
+		// continue with the rest of the steps.
+		//t.Fail()
```
Unused code?

joamaki replied:
Removed.

joamaki committed:
service: Rename Name to ServiceName and move to loadbalancer pkg

Name and namespace are now passed all the way to LBMap and we
already had the 'Name' struct in pkg/service. Unify the handling
of names by moving it to loadbalancer package and using it throughout.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

aditighag reviewed (approved):
Nice work! 👏 👏

christarazi commented on test/control-plane/agent.go:
```diff
@@ -0,0 +1,119 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package controlplane
```
Nit: I noticed that when [other tests](https://github.com/joamaki/cilium/pull/16) want to import this package, the import statement will need to be a named import:

```
controlplane github.com/cilium/cilium/test/control-plane
```

That shouldn't be necessary if we rename the directory to just `controlplane`.

christarazi commented on test/control-plane/services/dual-stack/run.go:
```diff
@@ -0,0 +1,52 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package dualstack
```
Same here with the named import comment

nbusseneau reviewed (approved):
Love it.

christarazi reviewed (approved):
https://github.com/cilium/cilium/pull/20320#discussion_r922710201 can be a followup.
